### ISSUE TITLE
Question regarding multiple DBs/shards

### ISSUE THREAD

**Issue by egezonberisha** [skeema/skeema#74]
1. What is a recommended way to match single dir/table against multiple/different databases?
For example, in cases where one DB represents one customer, would be nice to only track one sql file/table across all DBs, because that table would be the same across all DBs. 
_Note: A database gets created after a customers joins our platform, so we need to track/let "skeema" know about it and keep "skeema" updated every time that happens_

2. When we have a solution for question 1 above, and we want to change a specific table, is there a way to throttle or apply slowly the alter operations in that case, for example there are 100+ DBs that contain tableA, they are all in the same cluster,  and you don't want to modify tableA in 100 DBs simultaneously because of the load implications.

3. Combining single file/table across multiple DB with tiers. In the scenario where there is one DB per customer, is there any way to deploy to a percentage of DBs first, let's say we have 1000 databases in a cluster, and 30% of DBs are in tierA, 30% are in tierB etc. Is there any good way to deploy to tierA first, then after few days to tierB and so on.

Lastly, the `skeema init` taking too long to run in clusters with a lot of databases and tables per database, this specifically seems to when trying to get foreign keys 
https://github.com/skeema/tengo/blob/master/instance.go#L921
would be nice if there was an option to ignore tracking certain changes, for example we don't have foreign keys at all, so this is taking a long time to run and the results will always be 0.

**Comment by evanelias** [comment]
(1) You can map a dir to multiple schemas using one of the advanced forms of the [schema option](https://github.com/skeema/skeema/blob/master/doc/options.md#schema), and/or map a dir to multiple mysql-server instances using one of the advanced forms of the [host](https://github.com/skeema/skeema/blob/master/doc/options.md#host) and/or [host-wrapper](https://github.com/skeema/skeema/blob/master/doc/options.md#host-wrapper) options. To use these, just manually edit the generated `.skeema` files *after* you've run `skeema init`.

The exact setup will depend on your sharding topology and service discovery -- in other words, how you lay out your DBs and how you track them in your infrastructure.

(2) and (3) The [first-only](https://github.com/skeema/skeema/blob/master/doc/options.md#first-only) option allows you to push just to the first mapped host and schema per dir. 

For more fine-grained control, you could define a canary environment that only returns a subset of the full production environment's hosts/schemas. For example, perhaps you want `skeema push production` to target everything, but `skeema push canary` to only push to a portion. If you are shelling out to a custom script for `host-wrapper` and/or `schema`, you could pass the environment name to the script (use `{ENVIRONMENT}` on the script's configured command-line to have Skeema pass the right value in), and then make the script just output a subset of all schema names if the environment is "canary". 

I should also note that Skeema also won't operate on multiple targets *concurrently* by default. You can enable concurrency across different mysql instances using [concurrent-instances](https://github.com/skeema/skeema/blob/master/doc/options.md#concurrent-instances). There isn't a similar option for enabling concurrency across schemas on the *same* mysql instance yet, but one could be added if there's sufficient user interest.

(4) That's surprising -- I would have hoped that query to be very fast if there are no rows in `information_schema.referential_constraints`. Then again, some of those `information_schema` tables are tricky to optimize and can result in unexpected filesystem scans. A few questions to help debug:

* What database server version/flavor are you running?
* Any rough ballpark on how long it is taking / how many schemas (databases) / how many tables per database?
* Would you mind manully checking on the execution time of a potentially-optimized query, substituting in one of your schema names in where it says YOUR_DB_NAME_HERE --

a) query that Skeema is currently using
```sql
SELECT   rc.constraint_name AS constraint_name, rc.table_name AS table_name,
         rc.update_rule AS update_rule, rc.delete_rule AS delete_rule,
         rc.referenced_table_name AS referenced_table_name,
         IF(rc.constraint_schema=rc.unique_constraint_schema, '', rc.unique_constraint_schema) AS referenced_schema,
         kcu.referenced_column_name AS referenced_column_name,
         CONCAT(kcu.constraint_schema, '.', kcu.table_name, '.', kcu.column_name) AS col_lookup_key
FROM     information_schema.referential_constraints rc
JOIN     information_schema.key_column_usage kcu ON kcu.constraint_name = rc.constraint_name AND
                                 kcu.constraint_schema = rc.constraint_schema AND
                                 kcu.referenced_column_name IS NOT NULL
WHERE    rc.constraint_schema = 'YOUR_DB_NAME_HERE'
ORDER BY BINARY rc.constraint_name, kcu.ordinal_position;
```

vs

b) hopefully optimized query 🤞  Note that the database name must be put into 2 places in this one
```sql
SELECT   rc.constraint_name AS constraint_name, rc.table_name AS table_name,
         rc.update_rule AS update_rule, rc.delete_rule AS delete_rule,
         rc.referenced_table_name AS referenced_table_name,
         IF(rc.constraint_schema=rc.unique_constraint_schema, '', rc.unique_constraint_schema) AS referenced_schema,
         kcu.referenced_column_name AS referenced_column_name,
         CONCAT(kcu.constraint_schema, '.', kcu.table_name, '.', kcu.column_name) AS col_lookup_key
FROM     information_schema.referential_constraints rc
JOIN     information_schema.key_column_usage kcu ON kcu.constraint_name = rc.constraint_name AND
                                 kcu.table_schema = 'YOUR_DB_NAME_HERE' AND
                                 kcu.referenced_column_name IS NOT NULL
WHERE    rc.constraint_schema = 'YOUR_DB_NAME_HERE'
ORDER BY BINARY rc.constraint_name, kcu.ordinal_position;
```

I expect that query b may be much faster, unless you're using mysql 8. But I'm curious how much. Thanks!

**Comment by egezonberisha** [comment]
Thank you for your response, I will be looking at answer above. However just test the optimized query and its a lot faster, almost instantaneous! 
Using Percona 5.7.25

**Comment by evanelias** [comment]
Awesome, that is excellent info -- thank you for testing the queries and for reporting the original issue! I've just committed a change to Skeema to use the new query. It might be 1 to 2 weeks until the next precompiled binary release, but in the meantime `go get -u github.com/skeema/skeema` can be used to build it from source, if you have a Golang compiler (v1.9+).

More info on the original query's problem in https://github.com/skeema/tengo/commit/05dd6e2 if you're curious.

**Comment by evanelias** [comment]
The performance fix has been included in [Skeema v1.2.4](https://github.com/skeema/skeema/releases/tag/v1.2.4), released today. Thank you again for reporting the original issue!

Closing this out, but please feel free to re-open if you have any follow-up questions regarding configuring multiple DBs/shards in Skeema.
